File: Cpp/0396_rotate_function.cpp
// https://leetcode.cn/problems/rotate-function/

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int maxRotateFunction(vector<int>& nums) {
        int f = 0, s = 0, n = nums.size();
        for (int i = 0; i < n; ++i) {
            f += i * nums[i];
            s += nums[i];
        }
        int res = f;
        for (int i = 1; i < n; ++i) {
            f += s - n * nums[n - i];
            res = max(res, f);
        }
        return res;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {4, 3, 2, 6};
    cout << sol.maxRotateFunction(nums) << '\n';
    return 0;
}

File: Python3/0396_rotate_function.py
# https://leetcode.cn/problems/rotate-function/

class Solution:
    def maxRotateFunction(self, nums: list[int]) -> int:
        f, s = 0, 0
        for i, num in enumerate(nums):
            f += i * num
            s += num
        res = f
        n = len(nums)
        for i in range(1, n):
            f = f + s - n * nums[n-i]
            res = max(res, f)
        return res